Add Schema Versioning document (#12156)
microsoft/FluidFramework: packages/dds/SchemaVersioning.md

        This means old code stays in use and needs to be able to collaborate (on same doc) with new code seamlessly, at least for some amount of time.
        This window will differ depending on app and how it’s deployed – from couple weeks for web applications, to couple years for something like win32 Office Word.

1.  Documents can be read-only for some users, thus in-place conversion is not possible for all clients.

# Requirements

The above assumptions imply the following requirements for changes in document schema:

1.  Service based format conversions should never be required.

1.  Applications (including Fluid Framework bits) must support reading all formats that have ever been written.

1.  Any changes in document schema must ship dark (default to not being used for writing) and be roll out (enabling new format) on a schedule controlled by the controlled the team controlling the application deployment schedule.

    Writing new formats can only be enabled by application only when application knows it is close to 100% of clients having new bits to collaborate in new schema (or it’s Ok for old clients to fail in some controlled way).

> 1.  When a new format is added, support for writing the previous format must available in the same version that has support for the new format.

Author: I hadn't realized the above actually entails this. I'd added a note indicating that to prevent future reader from having to wonder the same thing.

    However the case where updating the application does not solve the issue should be handled as well: ensure no data corruption can occur due to unexpected formats being encountered as well as that if data corruption did occur and made a format appear unrecognized, that telemetry could detect the issue.

-   Keep the contents of persisted data formats out of public APIs.

    Since APIs have different compatibility requirements than persisted data formats, mixing the two will result in reduced evolvability long term.
    This avoids the complexity related to supporting reading and writing multiple formats from cluttering the APIs.

-   Implement support for persisted formats in a self contained way, where the relevant code is clearly marked.

    A type which is persisted has very different evolvability than one that is not.
    For example a package that has a non-exported interface usually can do a rename refactoring for members of the interface, but if that interface is used for Json serialized objects which are persisted,

>     the change could easily break support for all existing documents and/or collaboration with all existing clients, and do so in a way that won't show up in code review since the persistance/format logic might be untouched.

Reviewer: ```suggestion
    the change could easily break support for all existing documents and/or collaboration with all existing clients, and do so in a way that won't show up in code review since the persistence/format logic might be untouched.
```

>     Instead any code which defined a persisted format (including interfaces used with Json) should be clearly documented as such.

Reviewer: ```suggestion
    Instead any code which defines a persisted format (including interfaces used with Json) should be clearly documented as such.
```

    An easy way to do this is put persisted related details in a clearly named file (for example persisted_formats.ts),

>     and do not import anything into that file which could change and result in a changed data format (for example don't import any types to include in json serialized data).

Reviewer: ```suggestion
    and not importing anything into that file which could change and result in a changed data format (for example don't import any types to include in Json serialized data).
```

    If necessary duplicate types, or move the source of truth to the persisted format code.

-   Factor legacy format reading code (code for reading formats old enough that writing them is not supported) into compatibility libraries which can be authored as converters.

    This approach ensures that the only unbounded code growth from having a very old codebase with many legacy formats (support for reading all old formats) is organized in a way that its complexity does not increase as more formats are added, and it has no maintainability impact on other parts of the code.

## Libraries

This includes Fluid Framework, as well other libraries consumed by Fluid applications.

-   Adding support for writing a new format can be done in a minor version, but it must not be used unless opted into by the application or by editing a document already using the new format.

    When adding a new format, the APIs to opt into that format should explicitly document the oldest version in which that format is supported, so that when applications consider opting into it, they can know if it is safe for them to do so.

>     If the format is not yet stable (current version of it will not be supported for reading in all future versions), it must be explicitly documented as such,

Reviewer: This is an area we totally messed up with legacy SharedTree: we iterated on the 1.1.1 format in code, with it documented as not stable for production yet. However, we didn't actually rev the version to 1.1.2 or equivalent upon committing to backwards-compatibility. This ended up resulting in corrupted documents for the ~0.01% of sessions which were active for over 3 months between the dark rollout time and enabling the write feature, since they were using client versions which *thought* they knew how to interpret 1.1.1, but it was one of the intermediate iterations of 1.1.1. Thankfully the corrupted documents could be recovered using a rollback feature at the storage layer, but without that we would have needed to implement a lot of complicated hotfixes if we wanted to recover their data.

Clearly we did not follow best practices in this example. I do think there is high value in having a documented process for iterating on a work-in-progress eventual write format in a way which allows breaking format changes before finalizing the format and might be worth writing it as an example here. E.g., if we had instead kept the version number something like "1.1.1-dev" while we didn't commit to back-compat and finalized it via "promotion" to 1.1.1, those corrupt documents would have never happened.

Reviewer: fyi--my more recent thinking in this area is that it's probably just better to use numbers as versions, and while working on support for a new version have explicit guards that attempts to read/write that version fail. Ex: while working on V2, in code have a `maxProductionVersion: 1` field, create codecs which assert when encountering production version >1, add an internal-only backdoor which allows writing tests using version 2. then when ready to commit to backcompat for version 2, bump `maxProductionVersion` to 2.

Author: I have added additional guidance about this, which I think is even more protective than having a maxProductionVersion.

    dropping support for a format needs to be communicated to users of the library well in advance (more so than a simple API break that has no rollout requirements).
    For Fluid Framework this means announcing the deprecation of the format as part of the major release before the one where it is removed.

## Applications

-   Applications should opt into writing new data formats only once support for those formats has been fully deployed. This includes enough testing to be confident they won't need to roll back to a version without the support.
-   To best support Fluids maximum 30 day session length, applications may want to include an extra 30 days after the last rollout of support for the format before starting the rollout of enabling the new format for writing.
-   When updating to a new major version of a library (such as Fluid), check the release notes for current or upcoming format changes, and schedule format updates accordingly.

# Complications

It’s worth discussing what happens if we do not follow these rules.
A couple examples:

## Deprecating DDS

> We decide to deprecate some DDS by eventually removing it from code base. Clients who rely on it will be stuck, and have one of these options:

Reviewer: ```suggestion
Say we decide to deprecate some DDS and eventually remove it from the code base.
Clients who rely on this DDS will be stuck, and have one of these options:
```

1. Stay forever on old versions of Fluid Framework.
2. Copy DDS code into their repo and continue to support it the way it’s described at the beginning of the document.

It’s a possibility, but it’s very inefficient if many customers depend on it, then I’d argue it’s better to push it to separate repo for partners to share cost of supporting it. The obvious next step – we continue to support it, which is not much different form having it in our repo under “legacy” bucket.

Note that this workflow does not work with other changes, like changes in snapshot format.

On-the fly converters

> In some distant future, we may offer converter workflow, where (for example) Directory DDS can be morphed transparently into SharedTree DDS on open / first edit (and API not exposing Directory DDS at all), including any trailing ops. Based on assumptions above, we will need to deploy such changes dark and eventually trigger conversion. Due to offline clients having local changes in old format, we will need to deal with merging ops in old and new format. Given that in general, better support for offline will require addition of 3-way merge, we are likely to rely on 3-way merge to deal with these changes (it simplifies certain aspects of conversion, but requires more expensive, but more generic merge tooling).

Reviewer: ```suggestion
In some distant future, we may offer a converter workflow, where (for example) a Directory DDS can be morphed transparently into a SharedTree DDS on open / first edit (and the API does not expose the Directory DDS at all), including any trailing ops.
Based on the assumptions above, we will need to deploy such changes dark and eventually trigger conversion.
Due to offline clients having local changes in the old format, we will need to deal with merging ops in the old and new formats.
Given that in general, better support for offline will require addition of 3-way merge, we are likely to rely on 3-way merge to deal with these changes (it simplifies certain aspects of conversion, but requires more expensive, but more generic merge tooling).
```

(or break more on commas if you want; hard to judge if the sentences are reasonable in this editor)

While we did not think through designs here, I think it’s safe to say that we will not have these flows in production for next 2+ years.

# Appendix

> Shared Tree resources:

Reviewer: maybe worth calling out "legacy shared tree"? no strong preference.

Author: We do link one of its docs: I have fixed up the link descriptions to make that more clear.
